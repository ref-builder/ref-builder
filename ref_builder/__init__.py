"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
