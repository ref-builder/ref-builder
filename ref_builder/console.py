from collections.abc import Iterator

import rich.console
from rich.table import Table

from ref_builder.models import OTUMinimal
from ref_builder.plan import SegmentRule
from ref_builder.resources import RepoOTU


def _render_taxonomy_id_link(taxid: int) -> str:
    return f"[link=https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={taxid}]{taxid}[/link]"


def _render_nucleotide_link(accession: str) -> str:
    return f"[link=https://www.ncbi.nlm.nih.gov/nuccore/{accession}]{accession}[/link]"


def print_otu(otu: RepoOTU) -> None:
    """Print the details for an OTU to the console.

    :param otu: The OTU to print.

    """
    console.print(f"[bold][underline]{otu.name}[/bold][/underline]")
    console.line()

    table = Table(
        box=None,
        show_header=False,
    )

    table.add_row("[bold]ACRONYM[/bold]", otu.acronym)
    table.add_row("[bold]ID[/bold]", str(otu.id))

    if otu.legacy_id:
        table.add_row("[bold]LEGACY ID[/bold]", otu.legacy_id)

    table.add_row("[bold]TAXID[/bold]", _render_taxonomy_id_link(otu.taxid))

    max_accession_length = max(
        len(str(sequence.accession))
        for isolate in otu.isolates
        for sequence in isolate.sequences
    )

    max_segment_name_length = max(
        len(str(segment.name)) for segment in otu.plan.segments
    )

    console.print(table)

    console.line()
    console.print("[bold]PLAN[/bold]")
    console.line()

    plan_table = Table(box=None)

    plan_table.add_column("NAME")
    plan_table.add_column("REQUIRED")
    plan_table.add_column("LENGTH")
    plan_table.add_column("TOLERANCE")
    plan_table.add_column("ID")

    if not otu.plan.monopartite:
        for segment in otu.plan.segments:
            plan_table.add_row(
                str(segment.name),
                "[red]Yes[/red]"
                if segment.required == SegmentRule.REQUIRED
                else "[grey]No[/grey]",
                str(segment.length),
                str(segment.length_tolerance),
                str(segment.id),
            )
    else:
        monopartite_segment = otu.plan.segments[0]

        segment_name = (
            monopartite_segment.name
            if monopartite_segment.name is not None
            else "Unnamed"
        )

        plan_table.add_row(
            segment_name,
            "[red]Yes[/red]",
            str(monopartite_segment.length),
            str(monopartite_segment.length_tolerance),
        )

    console.print(plan_table)

    console.line()
    console.print("[bold]ISOLATES[/bold]")

    for isolate in otu.isolates:
        console.line()
        console.print(str(isolate.name)) if isolate.name is not None else console.print(
            "[UNNAMED]",
        )
        console.line()

        isolate_table = Table(
            box=None,
        )

        isolate_table.add_column("ACCESSION", width=max_accession_length)
        isolate_table.add_column("LENGTH")
        isolate_table.add_column("SEGMENT", min_width=max_segment_name_length)
        isolate_table.add_column("DEFINITION")

        for sequence in sorted(isolate.sequences, key=lambda s: s.accession):
            isolate_table.add_row(
                _render_nucleotide_link(str(sequence.accession)),
                str(len(sequence.sequence)),
                str(otu.plan.get_segment_by_id(sequence.segment).name),
                sequence.definition,
            )

        console.print(isolate_table)


def print_otu_list(otus: Iterator[OTUMinimal]) -> None:
    """Print a list of OTUs to the console.

    :param otus: The OTUs to print.

    """
    table = Table(
        box=None,
    )

    table.add_column("NAME")
    table.add_column("ACRONYM")
    table.add_column("TAXID")
    table.add_column("ID")

    added_rows = False

    for otu in otus:
        table.add_row(
            otu.name,
            otu.acronym,
            str(otu.taxid),
            str(otu.id),
        )

        added_rows = True

    if added_rows:
        console.print(table)
    else:
        console.print("No OTUs found")


console = rich.console.Console()
