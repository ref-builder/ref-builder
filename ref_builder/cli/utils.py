import sys
import warnings
from uuid import UUID

import click

from ref_builder.errors import InvalidInputError, PartialIDConflictError
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.repo import Repo
from ref_builder.utils import OTUDeletedWarning

pass_repo = click.make_pass_decorator(Repo)

UUID_STRING_LENGTH = 36


def get_otu_from_identifier(repo: Repo, identifier: str) -> OTUBuilder:
    """Return an OTU id from the repo if identifier matches a single OTU.

    The identifier can either be a stringified UUIDv4, a truncated portion
    of a UUID, a NCBI Taxonomy ID or an acronym associated with the OTU.

    Raise a PartialIDConflictError if >1 OTU is found.

    :param repo: the repository to be searched.
    :param identifier: a non-UUID identifier.
        Can be an integer Taxonomy ID, acronym or truncated partial UUID.
    :return: the UUID of the OTU or ``None``
    """
    otu_id = None

    if len(identifier) == UUID_STRING_LENGTH:
        try:
            otu_id = UUID(identifier)
        except ValueError:
            otu_id = None

    elif identifier.isnumeric():
        try:
            taxid = int(identifier)
        except ValueError:
            pass
        else:
            otu_id = repo.get_otu_id_by_taxid(taxid)

    else:
        if (otu_id := repo.get_otu_id_by_acronym(identifier)) is not None:
            return repo.get_otu(otu_id)

        try:
            otu_id = repo.get_otu_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError as e:
            click.echo(
                e,
                err=True,
            )

    if otu_id is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    with warnings.catch_warnings(
        category=OTUDeletedWarning, record=True
    ) as warning_list:
        otu_ = repo.get_otu(otu_id)

    if otu_ is None:
        if warning_list:
            click.echo("OTU has been deleted.", err=True)

        else:
            click.echo("OTU not found.", err=True)

        sys.exit(1)

    return otu_


def get_otu_isolate_ids_from_identifier(repo: Repo, identifier: str) -> (UUID, UUID):
    """Return an isolate ID from the repo if identifier matches a single isolate.

    Handles cases where the isolate cannot be found.
    """
    isolate_id = None

    if len(identifier) == 32:
        try:
            isolate_id = UUID(identifier)
        except ValueError:
            isolate_id = None

    else:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError:
            click.echo(
                "Partial ID segment must be at least 8 characters long.", err=True
            )
            sys.exit(1)

    if isolate_id is not None:
        if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is not None:
            if isolate_id in repo.get_otu(otu_id).isolate_ids:
                return otu_id, isolate_id

    click.echo("Isolate ID could not be found.", err=True)
    sys.exit(1)
