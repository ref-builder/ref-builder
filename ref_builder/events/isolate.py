from pydantic import UUID4

from ref_builder.events.base import EventData, Event, IsolateQuery
from ref_builder.utils import IsolateName


class CreateIsolateData(EventData):
    """The data for a :class:`CreateIsolate` event."""

    id: UUID4
    legacy_id: str | None
    name: IsolateName | None


class CreateIsolate(Event):
    """An event that creates an isolate for a specific OTU."""

    data: CreateIsolateData
    query: IsolateQuery


class DeleteIsolateData(EventData):
    """The data for a :class:`DeleteIsolate` event."""

    rationale: str


class DeleteIsolate(Event):
    """An isolate deletion event."""

    data: DeleteIsolateData
    query: IsolateQuery
