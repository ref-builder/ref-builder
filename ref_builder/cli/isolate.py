import sys
from pathlib import Path
from uuid import UUID

import click

from ref_builder.cli.utils import pass_repo
from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.isolate import (
    add_and_name_isolate,
    add_genbank_isolate,
    add_unnamed_isolate,
)
from ref_builder.otu.modify import delete_isolate_from_otu
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.repo import Repo, locked_repo
from ref_builder.utils import IsolateName, IsolateNameType


@click.group(name="isolate")
@click.pass_context
@path_option
def isolate(ctx: click.Context, path: Path) -> None:
    """Manage isolates."""
    ctx.obj = ctx.with_resource(locked_repo(path))


@isolate.command(name="create")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--taxid", type=int, required=True)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@pass_repo
def isolate_create(
    repo: Repo,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    taxid: int,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

        sys.exit(0)

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
            sys.exit(0)

        except RefSeqConflictError as e:
            click.echo(
                f"{e.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as e:
        click.echo(
            f"{e.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)


@isolate.command(name="delete")  # type: ignore
@click.argument("IDENTIFIER", type=str)
@pass_repo
def isolate_delete(repo: Repo, identifier: str) -> None:
    """Delete an isolate with a UUID corresponding to IDENTIFIER.

    IDENTIFIER is an unique isolate ID (>8 characters)
    """
    isolate_id = None
    try:
        isolate_id = UUID(identifier)
    except ValueError:
        pass

    if isolate_id is None:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except InvalidInputError:
            click.echo("Partial ID segment must be at least 8 characters long.", err=True)
            sys.exit(1)

        except PartialIDConflictError as e:
            click.echo(e, err=True)

    if isolate_id is None:
        click.echo("Isolate could not be found.", err=True)
        sys.exit(1)

    if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is None:
        click.echo(f"The containing OTU could not be found.", err=True)
        sys.exit(1)

    if (otu_ := repo.get_otu(otu_id)) is None:
        click.echo(f"OTU not found.", err=True)
        sys.exit(1)

    delete_isolate_from_otu(repo, otu_, isolate_id)

    click.echo("Isolate deleted.")
