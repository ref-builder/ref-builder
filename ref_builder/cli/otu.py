import sys
from pathlib import Path
from uuid import UUID

import click
from click import Context
import structlog

from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.console import print_otu, print_otu_list
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.isolate import (
    add_unnamed_isolate,
    add_and_name_isolate,
    add_genbank_isolate,
)
from ref_builder.otu.modify import (
    exclude_accessions_from_otu,
    set_representative_isolate,
    add_segments_to_plan,
    resize_monopartite_plan,
)
from ref_builder.otu.update import auto_update_otu, promote_otu_accessions
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.plan import SegmentRule, SegmentName
from ref_builder.repo import Repo
from ref_builder.utils import IsolateNameType, IsolateName

logger = structlog.get_logger()


@click.group(name="otu")
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@ignore_cache_option
@path_option
def otu_create(
    accessions_: list[str],
    acronym: str,
    ignore_cache: bool,
    path: Path,
    taxid: int,
) -> None:
    """Create a new OTU.

    OTUs are created from a list of accessions and a taxonomy ID. The associated Genbank
    records are fetched and used to create the first isolate and a plan.
    """
    repo = Repo(path)

    if len(accessions_) != len(set(accessions_)):
        click.echo("Duplicate accessions were provided.", err=True)
        sys.exit(1)

    try:
        create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> None:
    """Get an OTU by its unique ID or taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    print_otu(otu_)


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_minimal_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx: Context, path: Path, taxid: int) -> None:
    """Update the specified OTU with new data."""
    repo = Repo(path)

    ctx.ensure_object(dict)
    ctx.obj = {
        "REPO": repo,
        "TAXID": taxid,
    }

    if not repo.get_otu_id_by_taxid(taxid):
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_auto_update(ctx: Context, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="promote")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_promote_accessions(
    ctx: Context,
    ignore_cache: bool,
) -> None:
    """Promote all RefSeq accessions within this OTU."""
    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(ctx.obj["TAXID"])

    promote_otu_accessions(repo, otu_, ignore_cache)


@update.command(name="isolate")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@click.pass_context
def isolate_create(
    ctx: Context,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
        except RefSeqConflictError as err:
            click.echo(
                f"{err.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as err:
        click.echo(
            f"{err.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)

    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@update.command(name="exclude")  # type: ignore
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.pass_context
def accession_exclude(
    ctx: Context,
    accessions_: list[str],
) -> None:
    """Exclude the given accessions from this OTU.

    Any duplicate accessions will be ignored. Only one exclusion per unique accession
    will be made.
    """
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    exclude_accessions_from_otu(repo, otu_, accessions_)


@update.command(name="default")  # type: ignore
@click.argument("ISOLATE_KEY", type=str)
@click.pass_context
def otu_set_representative_isolate(
    ctx: Context,
    isolate_key: str,
) -> None:
    """Update the OTU with a new representative isolate."""
    taxid = ctx.obj["TAXID"]

    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        sys.exit(1)

    try:
        isolate_id = UUID(isolate_key)
    except ValueError:
        parts = isolate_key.split(" ")
        try:
            isolate_name = IsolateName(IsolateNameType(parts[0].lower()), parts[1])
        except ValueError:
            click.echo(f'Error: "{isolate_key}" is not a valid isolate name.', err=True)
            sys.exit(1)

        isolate_id = otu_.get_isolate_id_by_name(isolate_name)

    if isolate_id is None:
        click.echo("Isolate could not be found in this OTU.", err=True)
        sys.exit(1)

    set_representative_isolate(repo, otu_, isolate_id)


@update.group()
@click.pass_context
def plan(ctx: Context) -> None:
    """Add to and replace isolate plans for this OTU."""


@plan.command(name="extend")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_extend_segment_list(
    ctx: Context,
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Add recommended or optional segments to the OTU plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_segments_to_plan(
            repo,
            otu_,
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)


@plan.command(name="multipartite")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--name",
    type=(str, str),
    help='a name for the originalsegment, e.g. "RNA A"',
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_resize_monopartite(
    ctx: Context,
    name: tuple[str, str],
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Change a monopartite OTU plan to a multipartite plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    prefix, key = name

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        resize_monopartite_plan(
            repo,
            otu_,
            name=SegmentName(prefix=prefix, key=key),
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
