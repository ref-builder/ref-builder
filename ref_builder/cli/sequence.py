import sys
from pathlib import Path

import click

from ref_builder.cli.options import path_option
from ref_builder.cli.utils import get_otu_sequence_ids_from_identifier, pass_repo
from ref_builder.console import print_sequence, print_sequence_as_json
from ref_builder.repo import Repo, locked_repo


@click.group(name="sequence")
@click.pass_context
@path_option
def sequence(ctx: click.Context, path: Path) -> None:
    """Manage sequences."""
    ctx.obj = ctx.with_resource(locked_repo(path))

@sequence.command(name="get")
@click.argument("IDENTIFIER", type=str)
@click.option("--id-only", is_flag=True, help="Only output sequence ID.")
@click.option(
    "--json",
    "json_",
    is_flag=True,
    help="Output in JSON form.",
)
@pass_repo
def sequence_get(repo: Repo, identifier: str, id_only: bool, json_: bool) -> None:
    """Get an isolate with a UUID corresponding to IDENTIFIER.

    IDENTIFIER can be an accession number or a unique isolate ID (>8 characters)
    """
    otu_id, sequence_id = get_otu_sequence_ids_from_identifier(repo, identifier)

    otu_ = repo.get_otu(otu_id)

    sequence_ = otu_.get_sequence_by_id(sequence_id)

    if sequence_ is None:
        click.echo("Sequence could not be found.", err=True)
        sys.exit(1)

    if id_only:
        click.echo(sequence_.id)

        sys.exit(0)

    if json_:
        print_sequence_as_json(sequence_)

    else:
        print_sequence(sequence_)
