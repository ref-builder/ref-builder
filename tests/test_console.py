import pytest
from faker import Faker
from syrupy import SnapshotAssertion

from ref_builder.console import (
    console,
    print_isolate,
    print_isolate_as_json,
    print_otu,
    print_otu_as_json,
    print_otu_list,
)
from ref_builder.models import Molecule, MolType, OTUMinimal, Strandedness, Topology
from ref_builder.otu.builders.isolate import IsolateBuilder
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.otu.builders.sequence import SequenceBuilder
from ref_builder.plan import Plan, Segment, SegmentName, SegmentRule
from ref_builder.utils import Accession, IsolateName, IsolateNameType
from tests.fixtures.factories import IsolateFactory, OTUMinimalFactory
from tests.fixtures.providers import AccessionProvider, SequenceProvider


class TestPrintOTUList:
    """Tests for the ``print_otu_list`` function."""

    def test_ok(
        self,
        capsys: pytest.CaptureFixture,
        otu_minimal_factory: OTUMinimalFactory,
        snapshot: SnapshotAssertion,
    ) -> None:
        """Test that listed OTUs are printed."""
        print_otu_list(otu_minimal_factory.build() for _ in range(5))

        assert capsys.readouterr().out == snapshot

    def test_empty(self) -> None:
        """Test that an empty list of OTUs is printed."""
        with console.capture() as capture:
            print_otu_list(OTUMinimal(**otu) for otu in [])

        assert capture.get() == "No OTUs found\n"


class TestPrintIsolate:
    """Test isolate console output."""

    def test_ok(self, snapshot: SnapshotAssertion):
        """Test that an isolate is printed as expected by ``print_isolate``."""
        fake = Faker(["en_US"])
        fake.add_provider(AccessionProvider)
        fake.add_provider(SequenceProvider)
        fake.seed_instance(8801)

        plan = Plan.new(
            [
                Segment(
                    id=fake.uuid4(),
                    length=1099,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "R"),
                    rule=SegmentRule.REQUIRED,
                ),
                Segment(
                    id=fake.uuid4(),
                    length=1074,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "M"),
                    rule=SegmentRule.REQUIRED,
                ),
                Segment(
                    id=fake.uuid4(),
                    length=1087,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "S"),
                    rule=SegmentRule.REQUIRED,
                ),
            ],
        )

        isolate = IsolateBuilder(**IsolateFactory.build_on_plan(plan).model_dump())

        with console.capture() as capture:
            print_isolate(isolate, plan)

        assert capture.get() == snapshot

    def test_json_ok(self, snapshot: SnapshotAssertion):
        """Test that an isolate is printed as expected by ``print_isolate_as_json``."""
        fake = Faker(["en_US"])
        fake.add_provider(AccessionProvider)
        fake.add_provider(SequenceProvider)
        fake.seed_instance(8801)

        plan = Plan.new(
            [
                Segment(
                    id=fake.uuid4(),
                    length=1099,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "R"),
                    rule=SegmentRule.REQUIRED,
                ),
                Segment(
                    id=fake.uuid4(),
                    length=1074,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "M"),
                    rule=SegmentRule.REQUIRED,
                ),
                Segment(
                    id=fake.uuid4(),
                    length=1087,
                    length_tolerance=0.03,
                    name=SegmentName("DNA", "S"),
                    rule=SegmentRule.REQUIRED,
                ),
            ],
        )

        isolate = IsolateBuilder(**IsolateFactory.build_on_plan(plan).model_dump())

        with console.capture() as capture:
            print_isolate_as_json(isolate)

        assert capture.get() == snapshot


class TestPrintOTU:
    """Test OTU console output."""

    def test_print_otu(self, snapshot: SnapshotAssertion):
        """Test that an OTU is printed as expected by ``print_otu``."""
        fake = Faker(["en_US"])
        fake.add_provider(AccessionProvider)
        fake.add_provider(SequenceProvider)
        fake.seed_instance(8801)

        otu = OTUBuilder(
            id=fake.uuid4(),
            acronym="BabAV",
            excluded_accessions=set(),
            molecule=Molecule(
                strandedness=Strandedness.SINGLE,
                topology=Topology.LINEAR,
                type=MolType.RNA,
            ),
            name="Babuvirus abacae",
            plan=Plan.new(
                [
                    Segment(
                        id=fake.uuid4(),
                        length=1099,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "R"),
                        rule=SegmentRule.REQUIRED,
                    ),
                    Segment(
                        id=fake.uuid4(),
                        length=1074,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "M"),
                        rule=SegmentRule.REQUIRED,
                    ),
                    Segment(
                        id=fake.uuid4(),
                        length=1087,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "S"),
                        rule=SegmentRule.REQUIRED,
                    ),
                ],
            ),
            taxid=438782,
            isolates=[],
        )

        for _ in range(2):
            sequences = [
                SequenceBuilder(
                    id=fake.uuid4(),
                    accession=Accession.from_string(fake.accession() + ".1"),
                    definition=fake.sentence(),
                    sequence=fake.sequence(),
                    segment=segment.id,
                )
                for segment in otu.plan.segments
            ]

            otu.isolates.append(
                IsolateBuilder(
                    id=fake.uuid4(),
                    name=IsolateName(type=IsolateNameType.ISOLATE, value=fake.word()),
                    sequences=sequences,
                )
            )

        with console.capture() as capture:
            print_otu(otu)

        assert capture.get() == snapshot

    def test_as_json_ok(self, snapshot: SnapshotAssertion):
        """Test that an OTU is printed as expected by ``print_otu_as_json``."""
        fake = Faker(["en_US"])
        fake.add_provider(AccessionProvider)
        fake.add_provider(SequenceProvider)
        fake.seed_instance(8801)

        otu = OTUBuilder(
            id=fake.uuid4(),
            acronym="BabAV",
            excluded_accessions=set(),
            molecule=Molecule(
                strandedness=Strandedness.SINGLE,
                topology=Topology.LINEAR,
                type=MolType.RNA,
            ),
            name="Babuvirus abacae",
            plan=Plan(
                id=fake.uuid4(),
                segments=[
                    Segment(
                        id=fake.uuid4(),
                        length=1099,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "R"),
                        rule=SegmentRule.REQUIRED,
                    ),
                    Segment(
                        id=fake.uuid4(),
                        length=1074,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "M"),
                        rule=SegmentRule.REQUIRED,
                    ),
                    Segment(
                        id=fake.uuid4(),
                        length=1087,
                        length_tolerance=0.03,
                        name=SegmentName("DNA", "S"),
                        rule=SegmentRule.REQUIRED,
                    ),
                ],
            ),
            taxid=438782,
            isolates=[],
        )

        with console.capture() as capture:
            print_otu_as_json(otu)

        assert capture.get() == snapshot
