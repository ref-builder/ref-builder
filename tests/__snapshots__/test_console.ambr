# serializer version: 1
# name: TestPrintOTUList.test_ok
  '''
   NAME                           ACRONYM  TAXID   ID                             
   Sweet potato chlorotic virus   SPCV     109263  c963cfe0-afae-4a3b-b909-6a04e… 
   Sweet potato bunchy top virus  SPBTV    425603  f335c357-7972-436d-91b3-1a6c2… 
   Pepper flower break virus      PFBV     950682  3e1dcfb5-92bd-431c-b4d2-ea161… 
   Grapevine leaf crumple virus   GLCV     951539  3be4c78b-b4ae-4dd3-847e-60460… 
   Sweet potato leaf crumple      SPLCV    123000  4617edaa-a37f-4ba2-ac24-63432… 
   virus                                                                          
  
  '''
# ---
# name: test_print_otu
  '''
  Babuvirus abacae
  
   ACRONYM  BabAV                                
   ID       e7019832-f5b4-4b78-b528-9929bf0ec839 
   TAXID    438782                               
  
  SCHEMA
  
   SEGMENT  REQUIRED  LENGTH  TOLERANCE 
   DNA R    Yes       1099    0.03      
   DNA M    Yes       1074    0.03      
   DNA S    Yes       1087    0.03      
  
  ISOLATES
  
  Isolate organization
  
   ACCESSION    SEGMENT  DEFINITION                              
   I48166.1     DNA R    Stand audience who artist travel sell.  
   J64270.1     DNA S    Campaign college director less.         
   NC_167153.1  DNA M    Support wind good training main forget. 
  
  Isolate present
  
   ACCESSION    SEGMENT  DEFINITION                                      
   NC_132006.1  DNA M    Pull foot culture.                              
   NC_134696.1  DNA R    Poor beyond interest finally require wish most. 
   NC_795643.1  DNA S    Character game tell result marriage very.       
  
  '''
# ---
